File: unix/boot/spp/rpp/ratlibf/sctabl.f
      integer function sctabl (table, sym, info, posn)
      integer table, posn
      integer sym (100)
      integer info (100)
      integer mem( 60000)
      common/cdsmem/mem
      integer bucket, walker
      integer dsget
      integer nodsiz, i, j
      if (.not.(posn .eq. 0))goto 23000
      posn = dsget (2)
      mem (posn) = 1
      mem (posn + 1) = mem (table + 1)
23000 continue
      bucket = mem (posn)
      walker = mem (posn + 1)
      nodsiz = mem (table)
23002 continue
      if (.not.(walker .ne. 0))goto 23005
      i = walker + 1 + nodsiz
      j = 1
23007 if (.not.(mem (i) .ne. -2))goto 23008
      sym (j) = mem (i)
      i = i + 1
      j = j + 1
      goto 23007
23008 continue
      sym (j) = -2
      i = 1
23009 if (.not.(i .le. nodsiz))goto 23011
      j = walker + 1 + i - 1
      info (i) = mem (j)
23010 i = i + 1
      goto 23009
23011 continue
      mem (posn) = bucket
      mem (posn + 1) = mem (walker + 0)
      sctabl = 1
      return
23005 continue
      bucket = bucket + 1
      if (.not.(bucket .gt. 43))goto 23012
      goto 23004
23012 continue
      j = table + bucket
      walker = mem (j)
23006 continue
23003 goto 23002
23004 continue
      call dsfree (posn)
      posn = 0
      sctabl = -1
      return
      end

File: unix/boot/spp/rpp/ratlibr/sctabl.r
include	defs

# sctabl --- scan symbol table, returning next entry or EOF

   integer function sctabl (table, sym, info, posn)
   pointer table, posn
   character sym (ARB)
   integer info (ARB)

   DS_DECL(Mem, MEMSIZE)

   pointer bucket, walker
   pointer dsget

   integer nodsiz, i, j

   if (posn == 0) {		    # just starting scan?
      posn = dsget (2)		       # get space for position info
      Mem (posn) = 1		       # get index of first bucket
      Mem (posn + 1) = Mem (table + 1) # get pointer to first chain
      }

   bucket = Mem (posn)		    # recover previous position
   walker = Mem (posn + 1)
   nodsiz = Mem (table)

   repeat {    # until the next symbol, or none are left
      if (walker != LAMBDA) {	    # symbol available?
	 i = walker + ST_DATA + nodsiz
	 j = 1
	 while (Mem (i) != EOS) {
	    sym (j) = Mem (i)
	    i = i + 1
	    j = j + 1
	    }
	 sym (j) = EOS
	 for (i = 1; i <= nodsiz; i = i + 1) {
	    j = walker + ST_DATA + i - 1
	    info (i) = Mem (j)
	    }
	 Mem (posn) = bucket	    # save position of next symbol
	 Mem (posn + 1) = Mem (walker + ST_LINK)
	 sctabl = 1  # not EOF
	 return
	 }
      else {
	 bucket = bucket + 1
	 if (bucket > ST_HTABSIZE)
	    break
	 j = table + bucket
	 walker = Mem (j)
	 }
      }

   call dsfree (posn)	   # throw away position information
   posn = 0
   sctabl = EOF
   return
   end
